sqlglot/optimizer/eliminate_joins.py

```diff
@@ -0,0 +1,138 @@
+
+
+def eliminate_joins(expression):
+    """
+    Remove unused joins from an expression.
+
+    Currently, this only remove LEFT JOINs when we know that the join condition
+    doesn't produce duplicate rows.
+
+    Example:
+        >>> import sqlglot
+        >>> sql = "SELECT x.a FROM x LEFT JOIN (SELECT DISTINCT y.b FROM y) AS y ON x.b = y.b"
+        >>> expression = sqlglot.parse_one(sql)
+        >>> eliminate_joins(expression).sql()
+        'SELECT x.a FROM x'
+
+    Args:
+        expression (sqlglot.Expression): expression to optimize
+    Returns:
+        sqlglot.Expression: optimized expression
+    """
+    for scope in traverse_scope(expression):
+        # If any columns in this scope aren't qualified, it's hard to determine if a join isn't used.
+        # It's probably possible to infer this from the outputs of derived tables.
+        # But for now, let's just skip this rule.
+        if scope.unqualified_columns:
+            continue
+
+        joins = scope.expression.args.get("joins", [])
+
+        # Reverse the joins so we can remove chains of unused joins
+        for join in reversed(joins):
+            if join.side != "LEFT":
+                continue
+
+            on = join.args.get("on")
+            if not on:
+                continue
+
+            alias = join.this.alias_or_name
+
+            # We need to find all columns that reference this join.
+            # But columns in the ON clause shouldn't count.
+            on_clause_columns = set(id(column) for column in on.find_all(exp.Column))
+            join_is_used = any(column for column in scope.source_columns(alias) if id(column) not in on_clause_columns)
+            if join_is_used:
+                continue
+
+            # The join condition must include the entire set of unique outputs
+            inner_scope = scope.sources.get(alias)
+            unique_outputs = _unique_outputs(inner_scope)
+            if not unique_outputs:
+                continue
+            _, join_keys, _ = join_condition(join)
+            remaining_unique_outputs = unique_outputs - set(c.name for c in join_keys)
+            if remaining_unique_outputs:
+                continue
+
+            join.pop()
+            scope.remove_source(alias)
+    return expression
+
+
+def _unique_outputs(scope):
+    if scope.expression.args.get("distinct"):
+        return set(scope.expression.named_selects)
+
+    group = scope.expression.args.get("group")
+    if group:
+        grouped_expressions = set(group.expressions)
+        grouped_outputs = set()
+
+        unique_outputs = set()
+        for select in scope.selects:
+            output = select.unalias()
+            if output in grouped_expressions:
+                grouped_outputs.add(output)
+                unique_outputs.add(select.alias_or_name)
+
+        # All the grouped expressions must be in the output
+        if len(grouped_expressions - grouped_outputs) == 0:
```

Comment:
groped_expressions >= grouped_outptus

Reply:
This is a set operation. I'll use explicit set methods so the intent is clearer.